design-proposals/2022-10-27-rendering-modularization.md

```diff
@@ -0,0 +1,508 @@
+- Allow shaders to be instantiated by reference to unique names in pre-compiled file (Metal does this).
+- Set uniform values by name
+- Set uniform structs by name (Metal uses structs)
+- Tear down the specific shader (Program) definitions and replace them with one generic one.  This is almost the case now, with much of the key functionality being near the top of the class hierarchy.
+- Support for pure calculation shaders.  Useful for a variety of things, but think particles.
+
+#### Benefits
+We'll be able to:
+- Create entirely new shaders to do new and different things.
+- Replace existing shaders without having to recompile the whole set.
+- Let developers communicate with their own shaders.
+
+#### For GLES we’ll need to:
+- Push aside the existing Program hierarchy and rename it with a GLES extension.
+- Allow for named uniforms for new Programs<br>We can probably ignore that requirement for the existing shaders and just wrap them
+
+### Shader(Program) Registry
+
+_Addresses core functionality [#1](#core) and [#2](#core)._
+
+Once developers can create their own shaders, we’ll need a way of managing them, thus the Shader Registry.  This registry will associate a given shader (Program) with a name.  Basic functionality will include:
+- Add a new shader by name
+- Replace an existing shader by name
+- Query a shader by name
+- Publish the well known names of the core shaders for the basic functionality
+
+Layers already go through a level of indirection to instantiate or reference their associated programs.  The shader registry could be inserted into that get*LayerPrograms() pattern fairly easily.
+
+![Shader Registry](resources/fig1.png)
+
+The GLES version of the Shader Registry can encapsulate the simple (but fairly weird) logic of shader source lookup that exists now.
+
+#### The way it is now
+
+The layers more or less pull their shaders out of a compressed block of text.  Clever, but difficult to access.
+
+#### Required Changes
+
+We'll need to make a Shader Registry class, of course, but that's fairly simple.  It should be thread safe, but you're mostly going to be using it on the main thread.
+
+The layers should pull their shaders out of the shader on demand.  This will give developers a chance to replace or add shaders before they're needed.
+
+### Rendering Passes
+
+_Addresses core functionality [#5](#core)._
+
+Modern graphics pipelines use more than one pass to create a visual representation.  It's actually pretty unusual to have a single rendering pass in your toolkit.  These are used for a dizzying array of purposes, including lightning effects, shadows, atmosphere and many others.  We (Wet Dog Weather) use separate rendering passes to draw data intermingled with map elements.
+
+Without getting too deep into specifics, early rendering passes allow the developer to use the power of the rasterizer for their own data.  Later rendering passes are typically used to decorate the map with effects.
```

Comment:
I guess that we could (as a first step) pretend there is only a single rendering pass.
Maybe the toolkit only requires a single pass for now. Designing the first iteration with a single pass might simplify things.

Reply:
This is one of those changes it's easier to make once.  Multiple rendering passes isn't all that complicated when you're restructuring the rendering engine, but difficult to add from the outside.  I've done the latter for a proof of concept and it was annoying.